File: 04_arrays_part4/2_4sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        n = len(nums)
        nums.sort()
        for i in range(n):
            if i != 0 and nums[i] == nums[i - 1]:
                continue
            for j in range(i + 1, n):
                if j != i + 1 and nums[j] == nums[j - 1]:
                    continue
                left, right = j + 1, n - 1
                required_sum = target - nums[i] - nums[j]
                while left < right:
                    if nums[left] + nums[right] < required_sum:
                        left += 1
                    elif nums[left] + nums[right] > required_sum:
                        right -= 1
                    else:
                        res.append([nums[i], nums[j], nums[left], nums[right]])
                        left += 1
                        right -= 1
                        while left < right and nums[left] == nums[left - 1]:
                            left += 1
                        while left < right and nums[right] == nums[right + 1]:
                            right -= 1

        return res


class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = set()
        n = len(nums)
        nums.sort()
        for i in range(n):
            for j in range(i + 1, n):
                left, right = j + 1, n - 1
                required_sum = target - nums[i] - nums[j]
                while left < right:
                    if nums[left] + nums[right] < required_sum:
                        left += 1
                    elif nums[left] + nums[right] > required_sum:
                        right -= 1
                    else:
                        res.add((nums[i], nums[j], nums[left], nums[right]))
                        left += 1
                        right -= 1
        return res

File: 04_arrays_part4/2_4sum.rs
use std::collections::HashSet;

impl Solution {
    pub fn four_sum(nums: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut nums = nums;
        nums.sort();
        let n = nums.len();
        let mut res: HashSet<(i32, i32, i32, i32)> = HashSet::new();

        for i in 0..n {
            for j in i + 1..n {
                let (mut l, mut r) = (j + 1, n - 1);
                let req = target as i64 - nums[i] as i64 - nums[j] as i64;
                while l < r {
                    if (nums[l] + nums[r]) as i64 == req {
                        res.insert((nums[i], nums[j], nums[l], nums[r]));
                        l += 1;
                        r -= 1;
                    } else if ((nums[l] + nums[r]) as i64) < req {
                        l += 1;
                    } else {
                        r -= 1;
                    }
                }
            }
        }

        let mut return_value: Vec<Vec<i32>> = vec![];
        for (a, b, c, d) in res {
            return_value.push(vec![a, b, c, d]);
        }
        return_value
    }
}
